src/map.rs

```diff
@@ -1091,6 +1117,23 @@ where
         None
     }
 
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all pairs (k, v) such that f(&k,&v) returns false.
+    pub fn retain<F>(&mut self, mut f: F)
+        where
+            F: FnMut(&K, &V) -> bool
+    {
+        let guard = epoch::pin();
+        // removed selected keys
+        for (k, v) in self.iter(&guard) {
+            if !f(k, v) {
+                // TODO: right now this removing the element by key without checking the value. The value should be used somehow in old_value
+                self.replace_node(k, None, None, &guard);
```

Comment:
@jonhoo , Do we really need to check the old value here as in the java code does? I think I do not really understand that behavior:
If the value has change then why would I lose interest in removing it? I think if some thread commands to remove an entry should be ok, doesn't matter the value that entry holds. Or not?

Reply:
I believe the thinking is that if the old value has changed, then who are we to say that the user still wants to remove the element? I suppose we could also provide a method like `retain_force` (maybe pick a better name) that removes the value even if it has changed.

Reply:
Actually here [this](https://github.com/jonhoo/flurry/pull/36#discussion_r371968383) is needed also because of [this](https://github.com/jonhoo/flurry/blob/0c3bfb617be08279aaf54ddbe15cf22a9bed69ee/jsr166/src/ConcurrentHashMap.java#L1630), right?
So, the value here I have it as a `&V` already, but actually a [`Shared<'g, V>`](https://github.com/jonhoo/flurry/pull/36/files#diff-178822c448ce6afe0e383be6f514626eL1005) is what is used inside the replace method.
`Shared::from(self.get(k, &guard).unwrap())` makes the compiler cry :/

Reply:
What is the actual error you get? Does `Shared::from(... as *const _)` work?

Reply:
Sorry, should have report the error before:
`v` is of type `&V` already, taken from the iterator, so, when trying the "casting" into the `*const` I bump into this lifetime issue:
```
error[E0309]: the parameter type `V` may not live long enough
    --> src/map.rs:1134:33
     |
150  | impl<K, V, S> HashMap<K, V, S>
     |         - help: consider adding an explicit lifetime bound `V: 'g`...
...
1134 |                 let old_value : Shared<'g, V> = Shared::from(v as *const _);
     |                                 ^^^^^^^^^^^^^
     |
note: ...so that the type `V` will meet its required lifetime bounds
    --> src/map.rs:1134:33
     |
1134 |                 let old_value : Shared<'g, V> = Shared::from(v as *const _);
     |                                 ^^^^^^^^^^^^^

```

Reply:
Hmm, I thought the bound `V: 'g` would be inferred in the response from the iterator. It returns `&'g V`, right? Can you try being more explicit with `Shared::from(v as *const V)`?

Reply:
I found a way of making it compile:

```rust
impl<'g, K, V, S> HashMap<K, V, S>
where
    K: Sync + Send + Clone + Hash + Eq,
    V: Sync + Send + 'g,
    S: BuildHasher,
{

    /// Retains only the elements specified by the predicate.
    ///
    /// In other words, remove all pairs (k, v) such that f(&k,&v) returns false.
    pub fn retain<F>(&mut self, mut f: F)
        where
            F : FnMut(&K, &V) -> bool
    {
        let guard = epoch::pin();
        // removed selected keys
        for (k, v) in self.iter(&guard) {
            if !f(k, v) {
                // TODO: right now this removing the element by key without checking the value. The value should be used somehow in old_value
                let old_value : Shared<'g, V> = Shared::from(v as *const V);
                self.replace_node(
                    k,
                    None,
                    Some(old_value),
                    &guard,
                );
            }
        }
    }
}
```
I made and implicit `impl` for that live time, that solves the life time issues. The compiler error was actually saying so.

Reply:
Ohhh, wait, you actually had `Shared<'g, V>` in the code. Yeah, that probably won't work. I think what you want is just:
```rust
let old_value = Shared::from(v as *const V);
```
or _maybe_
```rust
let old_value: Shared<'_, V> = Shared::from(v as *const V);
```
The `'g` you've added to the `impl` block now is super weird, and I'm not sure how it even compiles since `'g` isn't bound to anything. When I wrote `'g`, what I mean was the lifetime of the `guard` that you create _inside_ the function.

Reply:
Aaaahhh, that was indeed the issue :/
I was using the specific lifetime and it wasn't even declared in the block.
Thanks! ⚙️